File: python/array_search/fibonacci_search_production.py
from dataclasses import dataclass


@dataclass(frozen=True)
class FibonacciSearchResult:
    found: bool
    index: int
    insertion_point: int


def fibonacci_lower_bound(values, target, *, key=None):
    key_fn = key or (lambda value: value)
    length = len(values)

    fib_mm2 = 0
    fib_mm1 = 1
    fib_m = 1

    while fib_m < length:
        fib_mm2 = fib_mm1
        fib_mm1 = fib_m
        fib_m = fib_mm2 + fib_mm1

    offset = -1

    while fib_m > 1:
        index = min(offset + fib_mm2, length - 1)

        if key_fn(values[index]) < target:
            fib_m = fib_mm1
            fib_mm1 = fib_mm2
            fib_mm2 = fib_m - fib_mm1
            offset = index
        else:
            fib_m = fib_mm2
            fib_mm1 = fib_mm1 - fib_mm2
            fib_mm2 = fib_m - fib_mm1

    return offset + 1


def fibonacci_search_first(values, target, *, key=None):
    index = fibonacci_lower_bound(values, target, key=key)
    key_fn = key or (lambda value: value)
    found = index < len(values) and key_fn(values[index]) == target

    return FibonacciSearchResult(
        found,
        index if found else -1,
        index
    )

File: java/array_search/FibonacciSearchProduction.java
import java.util.Comparator;
import java.util.List;

public final class FibonacciSearchProduction {
    public record SearchResult(boolean found, int index, int insertionPoint) {
    }

    private FibonacciSearchProduction() {
    }

    public static <T extends Comparable<? super T>> int fibonacciLowerBound(List<T> values, T target) {
        return fibonacciLowerBound(values, target, Comparator.naturalOrder());
    }

    public static <T> int fibonacciLowerBound(List<T> values, T target, Comparator<? super T> comparator) {
        if (values == null || comparator == null) {
            throw new IllegalArgumentException("values and comparator must not be null");
        }

        int length = values.size();
        int fibMm2 = 0;
        int fibMm1 = 1;
        int fibM = 1;

        while (fibM < length) {
            fibMm2 = fibMm1;
            fibMm1 = fibM;
            fibM = fibMm2 + fibMm1;
        }

        int offset = -1;

        while (fibM > 1) {
            int probe = Math.min(offset + fibMm2, length - 1);

            if (comparator.compare(values.get(probe), target) < 0) {
                fibM = fibMm1;
                fibMm1 = fibMm2;
                fibMm2 = fibM - fibMm1;
                offset = probe;
            } else {
                fibM = fibMm2;
                fibMm1 = fibMm1 - fibMm2;
                fibMm2 = fibM - fibMm1;
            }
        }

        return offset + 1;
    }

    public static <T extends Comparable<? super T>> SearchResult fibonacciSearchFirst(List<T> values, T target) {
        return fibonacciSearchFirst(values, target, Comparator.naturalOrder());
    }

    public static <T> SearchResult fibonacciSearchFirst(List<T> values, T target, Comparator<? super T> comparator) {
        int index = fibonacciLowerBound(values, target, comparator);
        boolean found = index < values.size() && comparator.compare(values.get(index), target) == 0;
        return new SearchResult(found, found ? index : -1, index);
    }
}
